### PR TITLE
fix(nextra): use `rehype-mdx-title` to determine page title

### PR DESCRIPTION
**Pull request by B2o5T**
fixes https://github.com/shuding/nextra/issues/514
<s>Note: this PR contains commits from PR https://github.com/shuding/nextra/pull/511 since I already touched `loader.ts` file. So merge https://github.com/shuding/nextra/pull/511 first and I'll rebase this PR</s>

- bump `vitest`
- fix unexpected behaviour of page title by installing `rehype-mdx-title` package

### CODE REVIEW

File: packages/nextra/__test__/fixture/headings/index.ts [hunk 1]
```diff
@@ -1,9 +0,0 @@
-import fs from 'fs/promises'
-import path from 'path'
-
-type Name = 'static-h1.mdx' | 'dynamic-h1.mdx' | 'no-h1.mdx'
-
-export async function loadFixture(name: Name) {
-  const filePath = path.join(process.cwd(), 'src', '__test__', 'fixture', 'headings', name)
-  return fs.readFile(filePath, 'utf8')
-}
```
Comment (B2o5T) [inline packages/nextra/__test__/fixture/headings/index.ts:9]: this file is nowhere unused. `compile.tests.ts` already contain `loadFixture` function

File: packages/nextra/src/loader.ts [hunk 2]
```diff
@@ -158,54 +151,56 @@ async function loader(
...
-  // Add imports and exports to the source
-  return prefix + '\n\n' + content + '\n\n' + suffix
+export default NextraPage
+`.trimStart()
 }
 
 export default function syncLoader(
   this: LoaderContext<LoaderOptions>,
   source: string,
-  callback: (err?: null | Error, content?: string | Buffer) => void
```
Comment (B2o5T) [inline packages/nextra/src/loader.ts:207]: current loader returns only string, but not Buffer

File: packages/nextra/src/mdx-plugins/remark.ts [hunk 3]
```diff
@@ -32,7 +30,7 @@ export function getFlattenedValue(node: Parent): string {
     .join('')
 }
 
-export default function remarkHeadings(this: Processor) {
+export function remarkHeadings(this: Processor) {
```
Comment (B2o5T) [inline packages/nextra/src/mdx-plugins/remark.ts:33]: it's better to use named export, as in `compile.ts` this function is imported as `remarkHandler`

File: packages/nextra/src/loader.ts [hunk 4]
```diff
@@ -158,54 +151,56 @@ async function loader(
...
 }
-NextraPage.getLayout = NextraLayout.getLayout`
 
-  // console.log(content)
+const NextraLayout = __nextra_withSSG__(__nextra_withLayout__({
+  filename: "${slash(filename)}",
+  route: "${slash(route)}",
+  meta: ${JSON.stringify(data)},
+  pageMap: __nextra_pageMap__,
+  titleText,
```
Comment (B2o5T) [inline packages/nextra/src/loader.ts:181]:
```suggestion
  titleText: typeof titleText === 'string' ? titleText : undefined,
```

File: packages/nextra/src/loader.ts [hunk 5]
```diff
@@ -203,8 +199,8 @@ export default NextraPage
 export default function syncLoader(
   this: LoaderContext<LoaderOptions>,
   source: string,
-  callback: (err?: null | Error, content?: string | Buffer) => void
-) {
+  callback: (err?: null | Error, content?: string) => void
```
Comment (B2o5T) [inline packages/nextra/src/loader.ts:202]:
```suggestion
  callback: (err?: null | Error, content?: string | Buffer) => void
```

File: packages/nextra-theme-blog/src/index.tsx [hunk 6]
```diff
@@ -186,7 +184,6 @@ const BlogLayout = ({
       <article className="container prose prose-sm md:prose dark:prose-dark">
         <HeadingContext.Provider value={ref}>
           {pageTitle ? <h1>{pageTitle}</h1> : null}
-          {mdxTitle ? <h1 ref={ref}></h1> : null}
```
[the remark this answers is not in the document]
Comment (shuding) [reply packages/nextra-theme-blog/src/index.tsx:189]: Legacy logic for rendering the title on the client side, should be removed anyways I think.

**Comment by shuding** [comment]
Thanks!

**Comment by shuding** [comment]
Heads up - I'm going to refactor the loader a bit because `rehype-mdx-title` adds an extra export to the source, that breaks [fast refresh](https://nextjs.org/docs/basic-features/fast-refresh#how-it-works) (only exports React component(s)).

**Comment by B2o5T** [comment]
@shuding if you add here `.replace("export const titleText", "const titleText")`, does it will fix your issue?

**Comment by shuding** [comment]
Yep that's how I fixed it locally 👍 @B2o5T

**Comment by B2o5T** [comment]
@shuding nice 😍

**Pull request merged**